File: src-tauri/sniffer_parser/src/serializable_packet/transport.rs
//! Transport level Packets Representation

use pnet::packet::icmp::echo_reply::EchoReplyPacket;
use pnet::packet::icmp::echo_request::EchoRequestPacket;
use pnet::packet::icmp::{IcmpPacket, IcmpType, IcmpTypes};
use pnet::packet::icmpv6::{Icmpv6Packet, Icmpv6Type, Icmpv6Types};
use pnet::packet::tcp::TcpPacket;
use pnet::packet::udp::UdpPacket;
use pnet::packet::Packet;
use serde::Serialize;

/// TCP Packet Representation
#[derive(Serialize, Debug, Clone)]
pub struct SerializableTcpPacket {
    pub source: u16,
    pub destination: u16,
    pub sequence: u32,
    pub acknowledgement: u32,
    pub data_offset: u8,
    pub reserved: u8,
    pub flags: u16,
    pub window: u16,
    pub checksum: u16,
    pub urgent_ptr: u16,
    pub options: Vec<u8>,
    pub length: usize,
}

impl<'a> From<&TcpPacket<'a>> for SerializableTcpPacket {
    fn from(packet: &TcpPacket<'a>) -> Self {
        SerializableTcpPacket {
            source: packet.get_source(),
            destination: packet.get_destination(),
            sequence: packet.get_sequence(),
            acknowledgement: packet.get_acknowledgement(),
            data_offset: packet.get_data_offset(),
            reserved: packet.get_reserved(),
            flags: packet.get_flags(),
            window: packet.get_window(),
            checksum: packet.get_checksum(),
            urgent_ptr: packet.get_urgent_ptr(),
            options: packet.get_options_raw().to_vec(),
            length: packet.payload().len(),
        }
    }
}

/// UDP Packet Representation
#[derive(Serialize, Debug, Clone)]
pub struct SerializableUdpPacket {
    pub source: u16,
    pub destination: u16,
    pub length: u16,
    pub checksum: u16,
}

impl<'a> From<&UdpPacket<'a>> for SerializableUdpPacket {
    fn from(packet: &UdpPacket<'a>) -> Self {
        SerializableUdpPacket {
            source: packet.get_source(),
            destination: packet.get_destination(),
            length: packet.get_length(),
            checksum: packet.get_checksum(),
        }
    }
}

/// ICMPv6 Packet Representation
#[derive(Serialize, Debug, Clone)]
pub struct SerializableIcmpv6Packet {
    pub icmpv6_type: String,
    pub icmpv6_code: u8,
    pub checksum: u16,
    pub length: usize,
}

impl<'a> From<&Icmpv6Packet<'a>> for SerializableIcmpv6Packet {
    fn from(packet: &Icmpv6Packet<'a>) -> Self {
        SerializableIcmpv6Packet {
            icmpv6_type: icmpv6_type_to_string(packet.get_icmpv6_type()),
            icmpv6_code: packet.get_icmpv6_code().0,
            checksum: packet.get_checksum(),
            length: packet.payload().len(),
        }
    }
}

/// Get ICMPv6 Message Type
pub fn icmpv6_type_to_string(icmp_type: Icmpv6Type) -> String {
    return match icmp_type {
        Icmpv6Types::DestinationUnreachable => format!("DestinationUnreachable ({})", icmp_type.0),
        Icmpv6Types::PacketTooBig => format!("PacketTooBig ({})", icmp_type.0),
        Icmpv6Types::TimeExceeded => format!("TimeExceeded ({})", icmp_type.0),
        Icmpv6Types::ParameterProblem => format!("ParameterProblem ({})", icmp_type.0),
        Icmpv6Types::EchoRequest => format!("EchoRequest ({})", icmp_type.0),
        Icmpv6Types::EchoReply => format!("EchoReply ({})", icmp_type.0),
        Icmpv6Types::RouterSolicit => format!("RouterSolicit ({})", icmp_type.0),
        Icmpv6Types::RouterAdvert => format!("RouterAdvert ({})", icmp_type.0),
        Icmpv6Types::NeighborSolicit => format!("NeighborSolicit ({})", icmp_type.0),
        Icmpv6Types::NeighborAdvert => format!("NeighborAdvert ({})", icmp_type.0),
        Icmpv6Types::Redirect => format!("Redirect ({})", icmp_type.0),
        _ => format!("Unknown ({})", icmp_type.0),
    };
}

/// ICMP Packet Representation
#[derive(Serialize, Debug, Clone)]
pub struct SerializableIcmpPacket {
    pub icmp_type: String,
    pub icmp_code: u8,
    pub checksum: u16,
    pub length: usize,
}

impl<'a> From<&IcmpPacket<'a>> for SerializableIcmpPacket {
    fn from(packet: &IcmpPacket<'a>) -> Self {
        SerializableIcmpPacket {
            icmp_type: icmp_type_to_string(packet.get_icmp_type()),
            icmp_code: packet.get_icmp_code().0,
            checksum: packet.get_checksum(),
            length: packet.payload().len(),
        }
    }
}

/// Get ICMPv4 Message Type
pub fn icmp_type_to_string(icmp_type: IcmpType) -> String {
    return match icmp_type {
        IcmpTypes::EchoReply => format!("EchoReply ({})", icmp_type.0),
        IcmpTypes::DestinationUnreachable => format!("DestinationUnreachable ({})", icmp_type.0),
        IcmpTypes::SourceQuench => format!("SourceQuench ({})", icmp_type.0),
        IcmpTypes::RedirectMessage => format!("RedirectMessage ({})", icmp_type.0),
        IcmpTypes::EchoRequest => format!("EchoRequest ({})", icmp_type.0),
        IcmpTypes::RouterAdvertisement => format!("RouterAdvertisement ({})", icmp_type.0),
        IcmpTypes::RouterSolicitation => format!("RouterSolicitation ({})", icmp_type.0),
        IcmpTypes::TimeExceeded => format!("TimeExceeded ({})", icmp_type.0),
        IcmpTypes::ParameterProblem => format!("ParameterProblem ({})", icmp_type.0),
        IcmpTypes::Timestamp => format!("Timestamp ({})", icmp_type.0),
        IcmpTypes::TimestampReply => format!("TimestampReply ({})", icmp_type.0),
        IcmpTypes::InformationRequest => format!("InformationRequest ({})", icmp_type.0),
        IcmpTypes::InformationReply => format!("InformationReply ({})", icmp_type.0),
        IcmpTypes::AddressMaskRequest => format!("AddressMaskRequest ({})", icmp_type.0),
        IcmpTypes::AddressMaskReply => format!("AddressMaskReply ({})", icmp_type.0),
        IcmpTypes::Traceroute => format!("Traceroute ({})", icmp_type.0),
        _ => format!("Unknown ({})", icmp_type.0),
    };
}

/// ICMP Echo Reply Packet Representation
#[derive(Serialize, Debug, Clone)]
pub struct SerializableEchoReplyPacket {
    pub icmp_type: u8,
    pub icmp_code: u8,
    pub checksum: u16,
    pub identifier: u16,
    pub sequence_number: u16,
    pub length: usize,
}

impl<'a> From<&EchoReplyPacket<'a>> for SerializableEchoReplyPacket {
    fn from(packet: &EchoReplyPacket<'a>) -> Self {
        SerializableEchoReplyPacket {
            icmp_type: packet.get_icmp_type().0,
            icmp_code: packet.get_icmp_code().0,
            checksum: packet.get_checksum(),
            identifier: packet.get_checksum(),
            sequence_number: packet.get_sequence_number(),
            length: packet.payload().len(),
        }
    }
}

/// ICMP Echo Request Packet Representation
#[derive(Serialize, Debug, Clone)]
pub struct SerializableEchoRequestPacket {
    pub icmp_type: u8,
    pub icmp_code: u8,
    pub checksum: u16,
    pub identifier: u16,
    pub sequence_number: u16,
    pub length: usize,
}

impl<'a> From<&EchoRequestPacket<'a>> for SerializableEchoRequestPacket {
    fn from(packet: &EchoRequestPacket<'a>) -> Self {
        SerializableEchoRequestPacket {
            icmp_type: packet.get_icmp_type().0,
            icmp_code: packet.get_icmp_code().0,
            checksum: packet.get_checksum(),
            identifier: packet.get_identifier(),
            sequence_number: packet.get_sequence_number(),
            length: packet.payload().len(),
        }
    }
}

File: src/serializable_packet/transport.ts
import {SerializableTransportLayerPacket} from "../types/sniffing";

/* rev */
const TCPflags = ["FIN", "SYN", "RST", "PSH", "ACK", "URG", "ECN", "CWR", "Nonce", "Reserved"]

export class TcpPacket implements SerializableTransportLayerPacket {
    source: number;
    destination: number;
    sequence: number;
    acknowledgement: number;
    data_offset: number;
    reserved: number;
    flags: number;
    window: number;
    checksum: number;
    urgent_ptr: number;
    options: [];
    payload: [];
    type: string;

    constructor(
        source: number,
        destination: number,
        sequence: number,
        acknowledgement: number,
        data_offset: number,
        reserved: number,
        flags: number,
        window: number,
        checksum: number,
        urgent_ptr: number,
        options: [],
        payload: [],
    ) {
        this.source = source;
        this.destination = destination;
        this.sequence = sequence;
        this.acknowledgement = acknowledgement;
        this.data_offset = data_offset;
        this.reserved = reserved;
        this.flags = flags;
        this.window = window;
        this.checksum = checksum;
        this.urgent_ptr = urgent_ptr;
        this.options = options;
        this.payload = payload;
        this.type = "Transmission Control Protocol"
    }

    getType(): string {
        return "TCP";
    }

    getInfo(): string {
        return this.source + " -> " + this.destination + " " + this.getFlags(this.flags) + " " +
            "Seq=" + this.sequence + " Ack=" + this.acknowledgement + " " +
            "Win=" + this.window + " Len=" + this.payload.length;
    }

    getFlags(flags: number): string {
        let f = flags.toString(2);

        let res = "[";

        for (let i = f.length - 1; i >= 0; i--) {
            if (f[i] === "1")
                res += TCPflags[i] + ",";
        }

        return res.slice(0, -1) + "]";
    }

    public toDisplay() {
        let packet_info = [];

        packet_info.push({"Destination Port": this.destination});
        packet_info.push({"Source Port": this.source});
        packet_info.push({"Sequence": this.sequence});
        packet_info.push({"Acknowledgment (ACK)": this.acknowledgement});
        packet_info.push({"Data Offset": this.data_offset});
        packet_info.push({"Reserved": this.reserved});
        packet_info.push({"Flags": this.flags});
        packet_info.push({"Windows": this.window});
        packet_info.push({"Checksum": this.checksum});
        packet_info.push({"Urgent Pointer": this.urgent_ptr});
        packet_info.push({"Options MAC": this.options});

        return packet_info;
    }

    public toString(): string {
        return this.type + ", Src Port: " + this.source + ", Dst Port: " + this.destination + ", Seq: " + this.sequence +
            ", Ack: " + this.acknowledgement
    }
}

export class UdpPacket implements SerializableTransportLayerPacket {
    source: number;
    destination: number;
    length: number;
    checksum: number;
    payload: [];
    type: string;

    constructor(
        source: number,
        destination: number,
        length: number,
        checksum: number,
        payload: []
    ) {
        this.source = source;
        this.destination = destination;
        this.length = length;
        this.checksum = checksum;
        this.payload = payload;
        this.type = "User Datagram Protocol"
    }

    public toDisplay() {
        let packet_info = [];

        packet_info.push({"Source": this.source});
        packet_info.push({"Destination": this.destination});
        packet_info.push({"Length": this.length});
        packet_info.push({"Checksum": this.checksum});

        return packet_info;
    }

    public toString(): string {
        return this.type + ", Src Port: " + this.source + ", Dst Port: " + this.destination
    }

    getInfo(): string {
        return this.source + " -> " + this.destination + " Len=" + this.payload.length;
    }

    getType(): string {
        return "UDP";
    }
}


export class Icmpv6Packet implements SerializableTransportLayerPacket {
    icmpv6_type: number;
    icmpv6_code: number;
    checksum: number;
    payload: [];
    type: string;

    constructor(
        icmpv6_type: number,
        icmpv6_code: number,
        checksum: number,
        payload: []
    ) {
        this.icmpv6_type = icmpv6_type;
        this.icmpv6_code = icmpv6_code;
        this.checksum = checksum;
        this.payload = payload;
        this.type = "Internet Control Message Protocol v6"
    }

    public toDisplay() {
        let packet_info = [];

        packet_info.push({"ICMP v6 Type": this.icmpv6_type});
        packet_info.push({"ICMP v6 Code": this.icmpv6_code});
        packet_info.push({"Checksum": this.checksum});

        return packet_info;
    }

    public toString(): string {
        return this.type
    }

    getInfo(): string {
        return ""; // TODO
    }

    getType(): string {
        return "ICMPv6";
    }
}

export class IcmpPacket implements SerializableTransportLayerPacket {
    icmp_type: number;
    icmp_code: number;
    checksum: number;
    payload: [];
    type: string;

    constructor(
        icmp_type: number,
        icmp_code: number,
        checksum: number,
        payload: []
    ) {
        this.icmp_type = icmp_type;
        this.icmp_code = icmp_code;
        this.checksum = checksum;
        this.payload = payload;
        this.type = "Internet Control Message Protocol"
    }

    public toDisplay() {
        let packet_info = [];

        packet_info.push({"ICMP Type": this.icmp_type});
        packet_info.push({"ICMP Code": this.icmp_code});
        packet_info.push({"Checksum": this.checksum});

        return packet_info;
    }

    public toString(): string {
        return this.type
    }

    getInfo(): string {
        return ""; // TODO
    }

    getType(): string {
        return "ICMP";
    }
}

export class EchoReply implements SerializableTransportLayerPacket {
    icmp_type: number;
    icmp_code: number;
    checksum: number;
    identifier: number;
    sequence_number: number;
    payload: [];
    type: string;

    constructor(
        icmp_type: number,
        icmp_code: number,
        checksum: number,
        identifier: number,
        sequence_number: number,
        payload: []
    ) {
        this.icmp_type = icmp_type;
        this.icmp_code = icmp_code;
        this.checksum = checksum;
        this.identifier = identifier;
        this.sequence_number = sequence_number;
        this.payload = payload;
        this.type = "Internet Control Message Protocol - Echo Reply";
    }

    public toDisplay() {
        let packet_info = [];

        packet_info.push({"ICMP Type": this.icmp_type});
        packet_info.push({"ICMP Code": this.icmp_code});
        packet_info.push({"Checksum": this.checksum});
        packet_info.push({"Identifier": this.identifier});
        packet_info.push({"Sequence Number": this.sequence_number});

        return packet_info;
    }

    public toString(): string {
        return this.type
    }

    getInfo(): string {
        return ""; // TODO
    }

    getType(): string {
        return ""; // TODO
    }
}

export class EchoRequest implements SerializableTransportLayerPacket {
    icmp_type: number;
    icmp_code: number;
    checksum: number;
    identifier: number;
    sequence_number: number;
    payload: [];
    type: string;

    constructor(
        icmp_type: number,
        icmp_code: number,
        checksum: number,
        identifier: number,
        sequence_number: number,
        payload: []
    ) {
        this.icmp_type = icmp_type;
        this.icmp_code = icmp_code;
        this.checksum = checksum;
        this.identifier = identifier;
        this.sequence_number = sequence_number;
        this.payload = payload;
        this.type = "Internet Control Message Protocol - Echo Request";
    }

    public toDisplay() {
        let packet_info = [];

        packet_info.push({"ICMP Type": this.icmp_type});
        packet_info.push({"ICMP Code": this.icmp_code});
        packet_info.push({"Checksum": this.checksum});
        packet_info.push({"Identifier": this.identifier});
        packet_info.push({"Sequence Number": this.sequence_number});

        return packet_info;
    }

    public toString(): string {
        return this.type
    }

    getInfo(): string {
        return ""; // TODO
    }

    getType(): string {
        return ""; // TODO
    }
}

